[TextFields] Make errorText span multiple lines on MDCTextField

andrewoverton opened the pull request:
Before:
![simulator screen shot - iphone 8 plus - 2018-07-09 at 13 58 27](https://user-images.githubusercontent.com/8020010/42467530-2e5787d2-8380-11e8-8cd6-e15931232b37.png)

After:
![simulator screen shot - iphone 8 plus - 2018-07-09 at 13 48 31](https://user-images.githubusercontent.com/8020010/42467469-ff4fd106-837f-11e8-8381-c4927ee2ea26.png)

Closes https://github.com/material-components/material-components-ios/issues/4476

andrewoverton committed:
Make errorText span multiple lines on MDCTextField

andrewoverton commented:
@willlarche, getting rid of the lines that set the `leadingUnderlineLabel`'s constraints' priorities to low caused `testConstraintPriorities` to fail... There's a note above that test case saying "All the constraints created internally by the MDCTextField need to have a rather low priority so they can be overriden by a controller from the outside". Is that still true? Couldn't a controller just set such constraints to be required, rather than high?

willlarche commented:
The input elements themselves are designed to lay their subviews out with lower priority constraints. That way a controller can always come in and override the constraints to give it a custom layout.

Looks like the most I ever thought that would be is 260. So, if you are changing constraints of the classes `MDCTextField` or `MDCMultilineTextField`, keep them less than 260 and when you make constraints from a controller, I like to start with 750 and go up. But required works too if your logic doesn't need any hierarchy of priority.

andrewoverton committed:
Make priorities high rather than required

andrewoverton commented:
Ah, you know what @willlarche , I mistakenly believed that constraint priorities were high by default, rather than required. So yep, my earlier comment doesn’t make sense. But what if we just set the constraints to be high here, as I did in 701ee35? That way an external controller would still be able to override them with required constraints, but the internal constraints would be prioritized enough to successfully lay out a multiline leading underline label.

Here is a screenshot with priorities at 260, the value you recommended I not exceed:
![simulator screen shot - iphone 8 plus - 2018-07-10 at 10 03 47](https://user-images.githubusercontent.com/8020010/42515403-b0fdf3c0-8429-11e8-8686-2549ca241ec4.png)

Here is a screenshot with priorities at 750/“default high”:
![simulator screen shot - iphone 8 plus - 2018-07-10 at 10 03 19](https://user-images.githubusercontent.com/8020010/42515392-ac13756a-8429-11e8-84d6-405fc4053999.png)

You can also see I set numberOfLines to be 2 instead of 0, as you told me offline that design said it shouldn't exceed two.

ivangodfather commented:
this will apply also for .helperText ? It has the same issue

willlarche commented:
@ivangodfather yes it will apply for helper text too!

willlarche reviewed (dismissed):
Spoke offline.

ianegordon commented:
@andrewoverton if this has been addressed offline, please merge.  If not please re-add Will to give approval when he gets back.
Many Thanks!

willlarche commented:
Then let’s close this. Andrew is working on changes.

andrewoverton commented:
Ok I'll close while I work on it 👍

andrewoverton committed:
Lower compressionResistancePriority for leadingUnderlineLabel

andrewoverton commented:
![simulator screen shot - iphone x - 2018-07-17 at 15 10 30](https://user-images.githubusercontent.com/8020010/42840050-9231f5f4-89d3-11e8-8a52-cc8c8f319158.png)

This is a screenshot with my most recent changes @willlarche :D

As we discussed, there is still this bug in the KitchenSink example where the TextFields' bounds change, both in this branch and develop. It is not yet clear whether or not this is a bug in the TextField itself or a consequence of the way this particular ViewController is built.

willlarche commented on components/TextFields/src/private/MDCTextInputCommonFundament.m:
```diff
@@ -384,14 +388,16 @@ - (void)setupUnderlineLabels {
   _trailingUnderlineLabel.opaque = NO;
   [_textInput addSubview:_trailingUnderlineLabel];
 
+  UILayoutPriority defaultPriority = UILayoutPriorityDefaultHigh;
```
We need to find a way to do this with constraints priorities below 260 in the input.

andrewoverton replied:
@willlarche, we prefer lower priorities because we need the controller to be able to override the constraints, correct? Can't a controller still override a constraint with a priority of 750 if it needs to?

It just seems to me that these constraints, which are there to keep the labels' contents inside the view bounds, will fail to do so if their priorities are too low and the text grows too long. That's actually what I would _expect_ to happen... 

When would a controller want to overpower these constraints, which seem to me to serve a vital purpose?
Would you be happier with an approach that adds fixed width constraints to the labels? I haven't tried it yet, but it might work as well...

Thank you for your attention/assistance!

apellegrino91 commented:
I have the same problem. How i can fix it?

willlarche commented:
@apellegrino91 , this is currently in development. You'll have to wait a few weeks for it to be resolved.

andrewoverton committed:
Reinstate low constraint priorities, set preferredMaxLayoutWidth, and base leading/trailing Y offsets on label intrinsic content size rather than font.lineHeight

andrewoverton commented:
Here are the latest changes @willlarche:

![simulator screen shot - iphone x - 2018-07-23 at 16 29 33](https://user-images.githubusercontent.com/8020010/43101460-7098ef16-8e96-11e8-9ab7-fad0a50362fc.png)

willlarche commented:
Found a couple bugs doing a quick run on my machine:

1. The trailing label has the wrong vertical placement. It should probably have its top pinned to the leading label if the leading label defines the vertical space.

![screen shot 2018-07-24 at 10 24 42 am](https://user-images.githubusercontent.com/1271525/43155335-fb93858c-8f2b-11e8-81a3-c8374ad13bc6.png)
![screen shot 2018-07-24 at 10 26 13 am](https://user-images.githubusercontent.com/1271525/43155344-028efc4a-8f2c-11e8-9ab8-371df10a9c59.png)


2. There's a lot of new ambiguous layout:

![screen shot 2018-07-24 at 10 24 46 am](https://user-images.githubusercontent.com/1271525/43155329-f72b61d6-8f2b-11e8-993b-d7fe7c811017.png)

andrewoverton committed:
Get rid of ambiguous layout warnings

andrewoverton committed:
Pin trailing label to underline rather than textInput bottom

andrewoverton commented:
@willlarche I addressed the ambiguous layout warnings and moved the trailing label to be pinned to the underline rather than the textInput bottom

willlarche commented:
Great! Found another bug.

Text areas are not returning the correct intrinsic content size. The error label's second line is out of bounds. Tested it with all the text inset modes.
![simulator screen shot - iphone x - 2018-07-26 at 16 03 23](https://user-images.githubusercontent.com/1271525/43293381-d9b57762-90ee-11e8-9d4c-0807717e6c53.png)

![screen shot 2018-07-26 at 4 03 14 pm](https://user-images.githubusercontent.com/1271525/43293152-993bcade-90ed-11e8-81df-8037ae3109c0.png)

willlarche commented:
Also, please add the ability to see multi-line errors into the kitchen sink controller and a test or two around the expected height.

andrewoverton committed:
Fix multiline error in outlined text area controller

andrewoverton committed:
Add ability to see multiline error message to kitchen sink example

andrewoverton committed:
Add test for multiline error messages

andrewoverton commented:
Omg! 😱I didn't notice that one. I just copied some of the textInset logic from the controller I had been working with to MDCTextInputControllerOutlinedTextArea. Here's a screenshot:
![simulator screen shot - iphone x - 2018-07-27 at 13 58 08](https://user-images.githubusercontent.com/8020010/43338690-3d950642-91a5-11e8-832c-e25df7b5931c.png)
That work is in 1541a94.

I made the requested changes to the kitchen sink example in 0a9bfc2.
And I added the test in a700b90. It's a little weird... I think it is doing what it's supposed to do though. It failed before 1541a94 and now it passes.

willlarche commented:
Oh, be sure to add some long helper text too, for testing. It gets this 2 line feature for free.
![simulator screen shot - iphone x - 2018-07-27 at 11 28 07](https://user-images.githubusercontent.com/1271525/43340150-3df624c8-9190-11e8-82a8-bfce1ecdf57b.png)

andrewoverton committed:
Add long helper text to kitchen sink example and try to fix spec that's failing in kokoro

willlarche commented:
Found another bug in the custom fonts example:
![simulator screen shot - iphone x - 2018-07-27 at 11 47 32](https://user-images.githubusercontent.com/1271525/43340984-fc95bd92-9192-11e8-85a8-1c6dc57b6ce3.png)
![screen shot 2018-07-27 at 11 48 06 am](https://user-images.githubusercontent.com/1271525/43340988-feddae5c-9192-11e8-8d7b-389cd074c036.png)

willlarche commented on components/TextFields/tests/unit/TextFieldTests.swift:
```diff
@@ -178,19 +178,27 @@ class TextFieldTests: XCTestCase {
     let controller1 = MDCTextInputControllerFilled(textInput: textField1)
     var errorMessage = "Error!"
     controller1.setErrorText(errorMessage, errorAccessibilityValue: nil)
+    textField1.setNeedsLayout()
```
Do you have to have this call? Is just `layoutIfNeeded` not enough?

willlarche commented:
Same in the filled example
![simulator screen shot - iphone x - 2018-07-27 at 11 53 48](https://user-images.githubusercontent.com/1271525/43341265-c56f6c2c-9193-11e8-8d14-0f7e7d8d8c76.png)

andrewoverton commented:
@willlarche , I didn't see that behavior in the custom font or filled example:

![simulator screen shot - iphone x - 2018-07-27 at 15 16 27](https://user-images.githubusercontent.com/8020010/43342267-22a804f0-91b0-11e8-8fe8-497cde2a4033.png)
![simulator screen shot - iphone x - 2018-07-27 at 15 12 07](https://user-images.githubusercontent.com/8020010/43342268-22b5e5d4-91b0-11e8-9b46-c0ee171a85bc.png)

andrewoverton committed:
Try to get specs passing

andrewoverton commented:
So bizarre... The tests were passing locally but failing in CI, so I chose one of the Xcode versions that was failing in CI (9.1), downloaded that version, got tests passing locally, pushed that up, and see now that it failed in CI again :/

andrewoverton committed:
Get rid of multiline text field multiline error text test case

jverkoey commented:
@andrewoverton  This PR is likely a good candidate for doing some smoke testing internally prior to landing in a release.

willlarche commented:
Also, I just realized we need to make sure you can opt-out of this and have only 1 line. Hopefully that's as easy as changing the number of lines the label can do.

willlarche commented on components/TextFields/src/private/MDCTextInputCommonFundament.m:
```diff
@@ -344,26 +348,37 @@ - (void)setupUnderlineLabels {
+                                  toItem:_trailingUnderlineLabel
+                               attribute:NSLayoutAttributeTop
+                              multiplier:1
+                                constant:-MDCTextInputHalfPadding];
```
I prefer a `-1 * [whatever]` for negating constants. It's easy to miss the hyphen by itself.

andrewoverton committed:
Use width constraint instead of preferredMaxLayoutWidth

andrewoverton commented:
Through a combination of setting the content compression resistance on the leading underline label AND adding a low priority width constraint I got truncation working with 1 line of text. That's in 350dc1b. I also made the default 1 line of text. That was a judgment call... I can set it back to two though.

andrewoverton committed:
Fix spec after making default number of lines 1

andrewoverton committed:
Make constraint constants more readable

willlarche commented:
Something seems to be inconsistent with the width constraint. This is what I see:

![simulator screen shot - iphone 8 plus - 2018-08-14 at 18 26 31](https://user-images.githubusercontent.com/1271525/44121956-b9eb3600-9fef-11e8-8395-1ed3ac54b84a.png)

andrewoverton committed:
Re-layout text field after call to updateLayout

andrewoverton commented:
@willlarche omg you'd posted a screenshot just like that 18 days ago and I fixed it and I guess I broke it in a commit since then? So finicky 😩Anyway, I assume what you're doing there is looping through the controllers in viewWillAppear and setting error text. I was just able to reproduce it doing that. Wasn't happening if I added the code in viewDidAppear, just viewWillAppear. Also, the layout would get fixed if I made the text field become the first responder. This makes it seem like it's just waiting on a layout pass that's not coming. Moving the setNeedsLayout/layoutSubviews call below the call to updateLayout in da1a6b7 fixes it.

Since it's already happened like 5 times on this PR I wouldn't be surprised if there are some other bugs lurking but I think we should just land this so I can deal with those as they may come. Something about having this thing sit around so long bothers me. Would you be okay with that?

willlarche commented:
No. If anything it’s proven this PR needs more rigorous testing. It’s far too dangerous at the moment. 

Let’s slow down and make sure everything is perfect.

andrewoverton commented:
Okay sounds good. Well hopefully da1a6b7 does the trick for now :)

jverkoey commented:
What's the status of this PR? If it's blocked please close it and re-open once it's ready to pursue again.

andrewoverton commented:
@willlarche and I talked about it and I think we're going to merge it into a feature branch and merge that into develop _eventually_ 😕

adrianod1as commented:
Is this (or will be) available in any branch?

romoore commented:
I believe this was never committed in develop. The work may be prioritized in the future, but at the moment it isn't at the top of our backlog.
